### ISSUE TITLE
question: How to use @ValidateNested decorator with groups?

### ISSUE THREAD

**Issue by fooddilsn** [typestack/class-validator#1402]
I'm trying to validate an object with some nested objects using different validation based on `groups`.

This is my (fake) class definition representing my actual code:

`
enum FooStatus {
  A = 'statusA',
  B = 'statusB',
  C = 'statusC',
}

class Amount {
  @IsNumberString()
  value: string;

  @IsIn(['EUR'])
  currency: string;
}

class Foo {
  @IsEnum(FooStatus)
  status: FooStatus;

  @Equals(undefined, {
    message: `amountA should not exist when status is one of the following values: ${FooStatus.A}, ${FooStatus.C}`,
    groups: [FooStatus.A, FooStatus.C],
  })
  @IsObject({ groups: [FooStatus.B] })
  @ValidateNested({ groups: [FooStatus.B] })
  @Type(() => Amount)
  amountA?: Amount;

  @Equals(undefined, {
    message: `amountB should not exist when status is one of the following values: ${FooStatus.A}, ${FooStatus.B}`,
    groups: [FooStatus.A, FooStatus.B],
  })
  @IsObject({ groups: [FooStatus.C] })
  @ValidateNested({ groups: [FooStatus.C] })
  @Type(() => Amount)
  amountB?: Amount;
}
`

and then I'm doing this:

`
const errors = await validate(plainToClass(Foo, payload), { groups: [payload.status] });
if (errors.length) {
  // throw my exception reporting error details
}
`

Decorators `@Equals` and `@IsObject` have the expected behaviour, but it seems like `@ValidateNested` is always ignored when I use `groups` option on it.

**Examples**:

`
// validation passes successfully (as expected)
{
  status: 'statusA'
}

// validation fails with message "amountA should not exist when status is ..." (as expected)
{
  status: 'statusA',
  amountA: 42
}

// validation fails with message "amountA must be an object" (as expected)
{
  status: 'statusB',
  amountA: 42
}

// validation passes successfully, but I expected that it fails due to the @ValidateNested decorator
{
  status: 'statusB',
  amountA: {
    invalid: true
  }
}
`

Is there a bug with `@ValidateNested` decorator?
Is it possible to use `@ValidateNested` decorator with `groups`?
Am I doing something wrong?

Thanks for helping me.

**Comment by anramanathan-fadv** [comment]
Any update on this question?

**Comment by sofiyanguyen** [comment]
In order for ValidateNested to work as expected with groups it requires the group to be added to the nested child property decorators - which is very cumbersome for deeply nested classes and seems like a bug with ValidateNested. Eg this works;
```
class Amount {
  @IsNumberString({ groups: [FooStatus.A, FooStatus.B, FooStatus.C ]})
  value: string;

  @IsIn(['EUR'], { groups: [FooStatus.A, FooStatus.B, FooStatus.C ]})
  currency: string;
}
```

**Comment by frostbytedata** [comment]
> In order for ValidateNested to work as expected with groups it requires the group to be added to the nested child property decorators - which is very cumbersome for deeply nested classes and seems like a bug with ValidateNested. Eg this works;
> 
> ```
> class Amount {
>   @IsNumberString({ groups: [FooStatus.A, FooStatus.B, FooStatus.C ]})
>   value: string;
> 
>   @IsIn(['EUR'], { groups: [FooStatus.A, FooStatus.B, FooStatus.C ]})
>   currency: string;
> }
> ```

I am having trouble with this as well. Are you saying the groups option should be present and identical on both the property in the parent class, as well as in the property of the nested class? Or just that the child property alone needs it?

**Comment by sofiyanguyen** [comment]
Yes, whichever group name you're filtering by needs to be present and identical on both the property in the parent class and the nested class (all the way down). This is only needed if you are validating while filtering on a group name though. In most cases I've found a way to validate the nested class separately and avoid the group name hack, eg,
`
var myFoo: Foo = new Foo();
const errors = Validate(myFoo.amountA);
`

**Comment by Sotatek-TuLe3** [comment]
Same issue, Do you have a solution?
